runtime/jnichk/jnichk_internal.h

```diff
@@ -102,9 +102,27 @@ typedef struct JNICHK_GREF_HASHENTRY {
 
 #if defined(J9VM_INTERP_ATOMIC_FREE_JNI)
 
-#define enterVM(currentThread) enterVMFromJNI(currentThread)
-
-#define exitVM(currentThread) exitVMToJNI(currentThread)
+#define enterVM(currentThread) \
+	BOOLEAN hasNoVMAccess = J9_ARE_NO_BITS_SET((currentThread)->publicFlags, J9_PUBLIC_FLAGS_VM_ACCESS); \
+	UDATA inNative = (currentThread)->inNative; \
+	do { \
+		if (inNative) { \
+			enterVMFromJNI(currentThread); \
+		} else { \
+			if (hasNoVMAccess) { \
+				acquireVMAccess(currentThread); \
+			} \
+		} \
+	} while(0)
+		
+#define exitVM(currentThread) \
+	do { \
+		if (inNative) { \
```

Comment:
Will `inNative` always be in-scope?  I'm assuming enterVM/exitVM are always used as a pair?

Reply:
Yes, the assumption is already there wrt the VM access variable.